### PR TITLE
interp: rewrite entire package 

### PR DESCRIPTION
**Pull request by aykevl**
For a full explanation, see interp/README.md. In short, this rewrite is a redesign of the partial evaluator which improves it over the previous partial evaluator. The main functional difference is that when interpreting a function, the interpretation can be rolled back when an unsupported instruction is encountered (for example, an actual unknown instruction or a branch on a value that's only known at runtime). This also means that it is no longer necessary to scan functions to see whether they can be interpreted: instead, this package now just tries to interpret it and reverts when it can't go further.

This new design has several benefits:

  * Most errors coming from the interp package are avoided, as it can simply skip the code it can't handle. This has long been an issue.
  * The memory model has been improved, which means some packages now pass all tests that previously didn't pass them.
  * Because of a better design, it is in fact a bit faster than the previous version.

This means the following packages now pass tests with `tinygo test`:

  * hash/adler32: previously it would hang in an infinite loop
  * math/cmplx: previously it resulted in errors

---

This PR depends on #1412, which should be merged first. Also, it requires a change to the go-llvm package.

### CODE REVIEW

**Comment by deadprogram** [comment]
#1412 has been merged. Not sure what change is needed to go-llvm or if that has already been made.

**Comment by aykevl** [comment]
No, a different change is needed. I'll get back to this, it really helps to make TinyGo a more correct Go compiler.

**Comment by aykevl** [comment]
Looks like this _almost_ works, but needs some work to support some Go versions.

**Comment by aykevl** [comment]
Note that this should probably be merged after the next release, to give it some time to find possible bugs.

**Comment by deadprogram** [comment]
Indeed :smiley_cat:

**Comment by aykevl** [comment]
Updated the PR, I think it is ready now. This newer version fixes an issue with the `append` builtin and (in a separate commit) fixes a bug with an assertion in the coroutine lowering pass (https://github.com/tinygo-org/tinygo/issues/1409 and https://github.com/tinygo-org/tinygo/issues/733).

**Comment by ofauchon** [comment]
With tinygo (dev branch), I could not compile my application: 

```
traceback:
/usr/lib/go/src/bytes/bytes.go:
  br i1 %14, label %for.body, label %for.done, !dbg !2460
/home/olivier/go/src/github.com/jacobsa/crypto/cmac/subkey.go:29:27:
  %9 = call { i8*, i32, i32 } @bytes.Repeat(i8* %6, i32 %7, i32 %8, i32 16, i8* undef, i8* undef), !dbg !2457
github.com/jacobsa/crypto/cmac/<init>:
  call void @"github.com/jacobsa/crypto/cmac.init#1"(i8* undef, i8* undef), !dbg !2452
```

Then I applied the 3 commits 

```
git cherry-pick b3794a9ea8af6b7c71112a15ad54470db060c66c
git cherry-pick 54ce74a784e05ac6c7d009aa840c802cf66fa1b6
git cherry-pick 0b3b692d533630e306bc815ac3387073bd954f8c
go install
```

This fixed the previous error ! 

Thanks 

Olivier

File: interp/interp.go [hunk 1]
```diff
@@ -1,59 +1,73 @@
...
-	name := "runtime.initAll"
-	e := &Eval{
-		Mod:          mod,
-		TargetData:   llvm.NewTargetData(mod.DataLayout()),
-		Debug:        debug,
-		dirtyGlobals: map[llvm.Value]struct{}{},
+	r := runner{
+		mod:           mod,
+		targetData:    llvm.NewTargetData(mod.DataLayout()),
+		debug:         debug,
```
Comment (niaow) [inline interp/interp.go:39]: Eventually we should standardize this type of functionality throughout the compiler, but this makes sense for now.

File: interp/memory.go [hunk 2]
```diff
@@ -0,0 +1,1392 @@
...
+		obj := mv.get(objectIndex)
+		if obj.marked < mark {
+			obj = obj.clone()
+			obj.marked = mark
+			if mv.objects == nil {
+				mv.objects = make(map[uint32]object)
+			}
+			mv.objects[objectIndex] = obj
+		}
+		// TODO: also mark objects referenced by this object.
```
Comment (niaow) [inline interp/memory.go]: Is this going to break anything now?

File: interp/memory.go [hunk 3]
```diff
@@ -0,0 +1,1392 @@
...
+func (v literalValue) asPointer(r *runner) (pointerValue, error) {
+	return pointerValue{}, errLiteralToPointer
+}
+
+func (v literalValue) asRawValue(r *runner) rawValue {
+	var buf []byte
+	switch value := v.value.(type) {
+	case uint64:
+		buf = make([]byte, 8)
+		binary.LittleEndian.PutUint64(buf, value)
```
Comment (niaow) [inline interp/memory.go:359]: I don't think we have to deal with any big-endian machines right now, but we should probably document this limitation somewhere.

**Review by niaow — CHANGES_REQUESTED** [review]
Looks good, one minor comment

File: interp/memory.go (hunk 3 above)
Comment (aykevl) [reply interp/memory.go:359]: Not sure whether this is necessary. If we ever add support for a big endian system (if they are still around), grepping for `LittleEndian` seems pretty obvious to me. Right now there is at least one other place that assumes little endian systems: builder/build.go (although it's perhaps less important as each architecture needs special code to be supported there).

File: interp/memory.go (hunk 2 above)
Comment (aykevl) [reply interp/memory.go]: In my testing, no. However it is technically unsound and should probably be fixed.

Comment (deadprogram) [reply interp/memory.go]: So, does this require additional work now, or should we merge and deal with in future effort?

File: interp/memory.go (hunk 3 above)
Comment (deadprogram) [reply interp/memory.go:359]: Makes sense to me!

Comment (aykevl) [reply interp/memory.go:359]: I will document this at the top of the file just to be sure.

File: interp/memory.go (hunk 2 above)
Comment (aykevl) [reply interp/memory.go]: Yes, I think I shouldn't be lazy and fix this (even if it doesn't seem to affect the behavior right now).

**Comment by aykevl** [comment]
Okay, I've written the missing bits (see the force push for a diff). Ready for review again.

File: interp/memory.go (hunk 2 above)
Comment (aykevl) [reply interp/memory.go]: Fixed now.

**Comment by deadprogram** [comment]
What is the status of this PR? It would appear that all feedback has now been addressed.

**Comment by aykevl** [comment]
Ready from my POV but maybe @niaow wants to comment.

**Review by niaow — APPROVED** [review]

**Comment by deadprogram** [comment]
Here we go!! :rocket: :raccoon:

**Comment by aykevl** [comment]
Unfortunately this appears to have broken at least support for the Circuit Playground Express. Even examples/blinky1 doesn't work anymore. ~It appears to be related to `machine.RingBuffer` (the `Put` and `Used` methods).~ I haven't investigated further yet.

**Comment by deadprogram** [comment]
OK that seems odd, because the TinyHCI tests passed on CPE board after the merge. 

However I am able right now to verify the problem on CPE myself. Also confirm it is commit 30df912565c2ab2d69efb40a2b2859a1736a88a3 that causes the problem.

**Comment by aykevl** [comment]
I'll take a look, but probably not before Saturday.

**Comment by deadprogram** [comment]
I wonder if it is specifically related to USB-CDC code somehow? 

I notice that now that this merged into `dev` in subsequent PRs that most of the ARM boards are failing, except for Microbit which along with RISC-V and AVR boards are still passing TinyHCI:
https://github.com/tinygo-org/tinygo/pull/1543/checks?check_run_id=1601942130

**Comment by aykevl** [comment]
Yes, probably. I will investigate soon.

**Comment by aykevl** [comment]
What I've discovered so far is that `-opt=1` seems to mask the issue. Not sure why or how yet, don't rely on it.

**Comment by aykevl** [comment]
I have figured out the issue, the fix is in #1547.

**Pull request merged**